{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "586a0a52-5dc8-49b5-a0a6-8e2d9b0bf95e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import csv\n",
    "import bioframe as bf\n",
    "import numpy as np\n",
    "from matplotlib import colors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dacf206d-9f56-4539-917c-291f7ffc48f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''Scan with Strand was through terminal - ask how to approach \n",
    "(either through terminal on notebook or using gimmemotifs Python API)\n",
    "Could potentially use snakemake file or bash script to work through this all'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0dd1ccb3-aacc-43ed-8264-6f7096011240",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input File\n",
    "df_scan = pd.read_csv('~/nakahashi_files/Nakahashicore.bed', sep='\\t', comment='#', \n",
    "                         names=['chrom', 'start', 'end', 'name', 'score', 'strand'])\n",
    "\n",
    "df_scan = df_scan.drop(columns=['name', 'score'])\n",
    "df_scan[0:3]\n",
    "\n",
    "#df_scan_up = bf.read_table('~/N_up.bed', schema='bed9', comment='#').keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8f802a29-cc34-4bd1-922c-6fa32ad829c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['chrom', 'start', 'end'], dtype='object')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "Index(['chrom', 'start', 'end'], dtype='object')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_1</th>\n",
       "      <th>start_1</th>\n",
       "      <th>end_1</th>\n",
       "      <th>chrom_2</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "      <td>chr1</td>\n",
       "      <td>8</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "      <td>chr1</td>\n",
       "      <td>12</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  chrom_1  start_1  end_1 chrom_2  start_2  end_2\n",
       "0    chr1        1      5    chr1        1      5\n",
       "1    chr1        1      5    chr1        3      8\n",
       "2    chr1        3      8    chr1        3      8\n",
       "3    chr1        8     10    chr1        8     10\n",
       "4    chr1       12     14    chr1       12     14\n",
       "5    chr1        3      8    chr1        1      5"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##Test Case\n",
    "df1 = pd.DataFrame([\n",
    "    ['chr1', 1, 5],\n",
    "    ['chr1', 3, 8],\n",
    "    ['chr1', 8, 10],\n",
    "    ['chr1', 12, 14]],\n",
    "    columns=['chrom', 'start', 'end']\n",
    ")\n",
    "display(df1.columns)\n",
    "\n",
    "df2 = pd.DataFrame([\n",
    "    ['chr1', 1, 5],\n",
    "    ['chr1', 3, 8],\n",
    "    ['chr1', 8, 10],\n",
    "    ['chr1', 12, 14]],\n",
    "    columns=['chrom', 'start', 'end']\n",
    ")\n",
    "display(df2.columns)\n",
    "\n",
    "import bioframe as bf\n",
    "\n",
    "overlapping_intervals = bf.overlap(df1, df2, how = 'inner', suffixes = ('_1', '_2'))\n",
    "display (overlapping_intervals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "85d9788e-f77f-42a3-a634-1a7686d180b3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_1</th>\n",
       "      <th>start_1</th>\n",
       "      <th>end_1</th>\n",
       "      <th>chrom_2</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "      <th>strand_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3005830</td>\n",
       "      <td>3005969</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3005860</td>\n",
       "      <td>3005864</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3039452</td>\n",
       "      <td>3039632</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3039475</td>\n",
       "      <td>3039479</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3280703</td>\n",
       "      <td>3280942</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3280856</td>\n",
       "      <td>3280860</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3337434</td>\n",
       "      <td>3337836</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3337697</td>\n",
       "      <td>3337701</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3495516</td>\n",
       "      <td>3495762</td>\n",
       "      <td>chr18</td>\n",
       "      <td>3495573</td>\n",
       "      <td>3495577</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38935</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98026545</td>\n",
       "      <td>98026859</td>\n",
       "      <td>chr16</td>\n",
       "      <td>98026597</td>\n",
       "      <td>98026601</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38936</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98100791</td>\n",
       "      <td>98101039</td>\n",
       "      <td>chr16</td>\n",
       "      <td>98100900</td>\n",
       "      <td>98100904</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38937</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98158558</td>\n",
       "      <td>98158780</td>\n",
       "      <td>chr16</td>\n",
       "      <td>98158689</td>\n",
       "      <td>98158693</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38938</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98243951</td>\n",
       "      <td>98244098</td>\n",
       "      <td>chr16</td>\n",
       "      <td>98243964</td>\n",
       "      <td>98243968</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38939</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98262861</td>\n",
       "      <td>98263092</td>\n",
       "      <td>chr16</td>\n",
       "      <td>98262946</td>\n",
       "      <td>98262950</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38940 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom_1   start_1     end_1 chrom_2   start_2     end_2 strand_2\n",
       "0       chr18   3005830   3005969   chr18   3005860   3005864        +\n",
       "1       chr18   3039452   3039632   chr18   3039475   3039479        +\n",
       "2       chr18   3280703   3280942   chr18   3280856   3280860        -\n",
       "3       chr18   3337434   3337836   chr18   3337697   3337701        +\n",
       "4       chr18   3495516   3495762   chr18   3495573   3495577        -\n",
       "...       ...       ...       ...     ...       ...       ...      ...\n",
       "38935   chr16  98026545  98026859   chr16  98026597  98026601        -\n",
       "38936   chr16  98100791  98101039   chr16  98100900  98100904        -\n",
       "38937   chr16  98158558  98158780   chr16  98158689  98158693        +\n",
       "38938   chr16  98243951  98244098   chr16  98243964  98243968        -\n",
       "38939   chr16  98262861  98263092   chr16  98262946  98262950        +\n",
       "\n",
       "[38940 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Combining DFs to identify overlaps\n",
    "overlapping_intervals = bf.overlap(df_Nakahashi, df_scan, how = 'inner', suffixes = ('_1', '_2'))\n",
    "display (overlapping_intervals)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ab27e50f-d21e-4fc9-b2a8-e1cd6f3ec5fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Splitting DFs into Pos and Neg Strands\n",
    "\n",
    "df_pos = overlapping_intervals[overlapping_intervals['strand_2'] == '+']\n",
    "df_neg = overlapping_intervals[overlapping_intervals['strand_2'] == '-']\n",
    "\n",
    "## 1 is positive-strand \n",
    "## 0 is negative-strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dafaaf51-ef7c-479c-bca0-e0e79bd878ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3005969</td>\n",
       "      <td>3006169</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3039632</td>\n",
       "      <td>3039832</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3337836</td>\n",
       "      <td>3338036</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom    start      end strand\n",
       "0  chr18  3005969  3006169      +\n",
       "1  chr18  3039632  3039832      +\n",
       "3  chr18  3337836  3338036      +"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def create_pos_downstream(df_pos):\n",
    "    df_downstream_1 = df_pos.copy()\n",
    "     #display(df_downstream_1)\n",
    "        \n",
    "    df_downstream_1['start_1'] = df_downstream_1['start_1'].astype(int)\n",
    "    df_downstream_1['end_1'] = df_downstream_1['end_1'].astype(int)\n",
    "    df_downstream_1['start_1'] = df_downstream_1['end_1']\n",
    "    df_downstream_1['end_1'] = df_downstream_1['start_1'] + 200\n",
    "\n",
    "    df_downstream_1 = df_downstream_1.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_downstream_1.columns.values[0] = '#chr'\n",
    "    df_downstream_1.columns.values[1] = 'start'\n",
    "    df_downstream_1.columns.values[2] = 'end'\n",
    "    df_downstream_1.to_csv('downstream_pos.bed', sep ='\\t', index = False)\n",
    "    downstream_pos = df_downstream_1 \n",
    "    return downstream_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3eb6c7be-0111-4d0c-9a5a-0714af5c440e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_pos_upstream(df_pos):\n",
    "    df_upstream_1 = df_pos.copy()\n",
    "    #display(df_upstream_1)\n",
    "    \n",
    "    df_upstream_1['start_1'] = df_upstream_1['start_1'].astype(int)\n",
    "    df_upstream_1['end_1'] = df_upstream_1['end_1'].astype(int)\n",
    "    df_upstream_1['end_1'] = df_upstream_1['start_1']\n",
    "    df_upstream_1['start_1'] = df_upstream_1['start_1'] - 200\n",
    "    \n",
    "    df_upstream_1 = df_upstream_1.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_upstream_1.columns.values[0] = '#chr'\n",
    "    df_upstream_1.columns.values[1] = 'start'\n",
    "    df_upstream_1.columns.values[2] = 'end'\n",
    "    df_upstream_1.to_csv('upstream_pos.bed', sep ='\\t', index = False)\n",
    "    upstream_pos = df_upstream_1 \n",
    "    return upstream_pos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a2f3d255-6aa2-42d7-ba0c-8dd285117b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_neg_upstream(df_neg):\n",
    "    df_upstream_0 = df_neg.copy()\n",
    "    #display(df_upstream_0)\n",
    "\n",
    "    df_upstream_0['start_1'] = df_upstream_0['start_1'].astype(int)\n",
    "    df_upstream_0['end_1'] = df_upstream_0['end_1'].astype(int)\n",
    "    df_upstream_0['end_1'] = df_upstream_0['start_1']\n",
    "    df_upstream_0['start_1'] = df_upstream_0['start_1'] - 200\n",
    "\n",
    "    df_upstream_0 = df_upstream_0.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_upstream_0.columns.values[0] = '#chr'\n",
    "    df_upstream_0.columns.values[1] = 'start'\n",
    "    df_upstream_0.columns.values[2] = 'end'\n",
    "    df_upstream_0.to_csv('upstream_neg.bed', sep ='\\t', index = False)\n",
    "    upstream_neg = df_upstream_0 \n",
    "    return upstream_neg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ef457c01-83e6-4fc5-b755-828f15756b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_neg_downstream(df_neg):\n",
    "    df_downstream_0 = df_neg.copy()\n",
    "    #display(df_downstream_0)\n",
    "    \n",
    "    df_downstream_0['start_1'] = df_downstream_0['start_1'].astype(int)\n",
    "    df_downstream_0['end_1'] = df_downstream_0['end_1'].astype(int)\n",
    "    df_downstream_0['start_1'] = df_downstream_0['end_1']\n",
    "    df_downstream_0['end_1'] = df_downstream_0['start_1'] + 200\n",
    "\n",
    "    df_downstream_0 = df_downstream_0.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_downstream_0.columns.values[0] = '#chr'\n",
    "    df_downstream_0.columns.values[1] = 'start'\n",
    "    df_downstream_0.columns.values[2] = 'end'\n",
    "    df_downstream_0.to_csv('downstream_neg.bed', sep ='\\t', index = False)\n",
    "    downstream_neg = df_downstream_0 \n",
    "    return downstream_neg"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "72a4a653-4b3e-4244-8972-37a7dd134af6",
   "metadata": {},
   "outputs": [],
   "source": [
    "## This code performs de novo motif discovery on the dataframes\n",
    "'''I have to switch conda environments - there are some dependencies that prevent me \n",
    "from having bioframe and gimmemotifs within the same env. '''\n",
    "\n",
    "from gimmemotifs.denovo import gimme_motifs\n",
    "from gimmemotifs.scanner import Scanner\n",
    "\n",
    "peaks = ['upstream_pos.bed', 'downstream_pos.bed', 'upstream_neg.bed', 'upstream_neg.bed']\n",
    "outdir = ['upstream_pos.bed', 'downstream_pos.bed', 'upstream_neg.bed', 'upstream_neg.bed']\n",
    "params = {\n",
    "    \"tools\": \"Homer,BioProspector,MEME\",\n",
    "    \"genome\": \"mm9\"\n",
    "    }\n",
    "\n",
    "for (x,y) in zip(peaks, outdir): \n",
    "    motifs = gimme_motifs(x, y, params=params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "4460ea04-b39f-41fb-82ee-029820600fa2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3005969</td>\n",
       "      <td>3006169</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3039632</td>\n",
       "      <td>3039832</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3337836</td>\n",
       "      <td>3338036</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3497634</td>\n",
       "      <td>3497834</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr18</td>\n",
       "      <td>3508038</td>\n",
       "      <td>3508238</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38932</th>\n",
       "      <td>chr16</td>\n",
       "      <td>97790546</td>\n",
       "      <td>97790746</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38933</th>\n",
       "      <td>chr16</td>\n",
       "      <td>97815065</td>\n",
       "      <td>97815265</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38934</th>\n",
       "      <td>chr16</td>\n",
       "      <td>97889463</td>\n",
       "      <td>97889663</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38937</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98158780</td>\n",
       "      <td>98158980</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38939</th>\n",
       "      <td>chr16</td>\n",
       "      <td>98263092</td>\n",
       "      <td>98263292</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28136 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom     start       end strand\n",
       "0      chr18   3005969   3006169      +\n",
       "1      chr18   3039632   3039832      +\n",
       "3      chr18   3337836   3338036      +\n",
       "5      chr18   3497634   3497834      +\n",
       "6      chr18   3508038   3508238      +\n",
       "...      ...       ...       ...    ...\n",
       "38932  chr16  97790546  97790746      +\n",
       "38933  chr16  97815065  97815265      +\n",
       "38934  chr16  97889463  97889663      +\n",
       "38937  chr16  98158780  98158980      +\n",
       "38939  chr16  98263092  98263292      +\n",
       "\n",
       "[28136 rows x 4 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "<class 'int'>\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "an integer is required",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/SLURM_7162758/ipykernel_20802/1775540702.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[0mdisplay\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msites\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m \u001b[0;31m#print(bf.is_bedframe(sites))\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 94\u001b[0;31m \u001b[0mheatmap\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msites\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/tmp/SLURM_7162758/ipykernel_20802/1775540702.py\u001b[0m in \u001b[0;36mheatmap\u001b[0;34m(sites)\u001b[0m\n\u001b[1;32m     61\u001b[0m         \u001b[0;31m#chrom, start, end, strand = sites.loc[[i]][['chrom','start','end','strand']]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m         \u001b[0;34m''' Error occuring here with genome_open.fetch - the program wants integers, but I am unable to cast  (Cast Variable Types) '''\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 63\u001b[0;31m         \u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mgenome_open\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfetch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mchrom\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstart\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     64\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mstrand\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'-'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     65\u001b[0m             \u001b[0mseq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdna_rc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpysam/libcfaidx.pyx\u001b[0m in \u001b[0;36mpysam.libcfaidx.FastaFile.fetch\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpysam/libcutils.pyx\u001b[0m in \u001b[0;36mpysam.libcutils.parse_region\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: an integer is required"
     ]
    }
   ],
   "source": [
    "## Visualization of Findings through Heatmap \n",
    "## two nice colormaps\n",
    "cmap = colors.ListedColormap([\n",
    "  'mistyrose', \n",
    "  'darkseagreen',\n",
    "  'plum',\n",
    "  'beige'])\n",
    "\n",
    "cmap_rhee = colors.ListedColormap([\n",
    "    'red', #a red\n",
    "    'blue', #c blue\n",
    "    'yellow', #g yllow\n",
    "    'green' #t green\n",
    "])\n",
    "\n",
    "## dna utilties\n",
    "def dna_rc(seq):\n",
    "  return seq.translate(str.maketrans(\"ATCGatcg\",\"TAGCtagc\"))[::-1]\n",
    "\n",
    "def dna_1hot_index(seq):\n",
    "  seq_len = len(seq)\n",
    "  seq = seq.upper()\n",
    "  # map nt's to a len(seq) of 0,1,2,3\n",
    "  seq_code = np.zeros(seq_len, dtype='uint8')\n",
    "  for i in range(seq_len):\n",
    "    nt = seq[i]\n",
    "    if nt == 'A':\n",
    "      seq_code[i] = 0\n",
    "    elif nt == 'C':\n",
    "      seq_code[i] = 1\n",
    "    elif nt == 'G':\n",
    "      seq_code[i] = 2\n",
    "    elif nt == 'T':\n",
    "      seq_code[i] = 3\n",
    "    else:\n",
    "      seq_code[i] = random.randint(0,3)         \n",
    "  return seq_code\n",
    "\n",
    "import pysam\n",
    "genome_open = pysam.Fastafile('/project/fudenber_735/genomes/mm10/mm10.fa')\n",
    "\n",
    "\n",
    "## assuming sites is a dataFrame of motif positions\n",
    "def heatmap(sites): \n",
    "    pad = 100\n",
    "    sites_dna_num = []\n",
    "    #sites = sites.astype({'start':int,'end':int'})\n",
    "    \n",
    "    \n",
    "    for i in range(len(sites)):\n",
    "        if np.mod(i, 100)==0: print(i)\n",
    "        chrom = sites.loc[[i]][['chrom']]\n",
    "        start = sites.loc[[i]][['start']].astype(int)\n",
    "        end = sites.loc[[i]][['end']].astype(int)\n",
    "        strand = sites.loc[[i]][['strand']]\n",
    "        start = start - pad\n",
    "        end = end + pad\n",
    "        print(type(pad))\n",
    "        print(type(start))\n",
    "        print(type(end))\n",
    "        #chrom, start, end, strand = sites.loc[[i]][['chrom','start','end','strand']]\n",
    "        ''' Error occuring here with genome_open.fetch - the program wants integers, but I am unable to cast them as integers \n",
    "        despite deliberately doing so  (Cast Variable Types) '''\n",
    "        seq =  genome_open.fetch(chrom, start, end).upper()\n",
    "        if strand == '-':\n",
    "            seq = dna_rc(seq)\n",
    "        sites_dna_num.append(dna_1hot_index(seq))\n",
    "    sites_dna_num = np.array(sites_dna_num)\n",
    "\n",
    "    ## optional: cluster by hamming distance between sequences in the 20bp upstream of the motif\n",
    "    ## and use this to re-order the heatmap of sequences.\n",
    "    sub_index = (pad-20,pad)\n",
    "    seq_dist = np.zeros((len(sites),len(sites)))\n",
    "    for i in range(len(sites)):\n",
    "        seq_i = sites_dna_num[i][sub_index[0]:sub_index[1]]\n",
    "        for j in range(len(sites_strong)):\n",
    "            if i<j:\n",
    "                seq_j = sites_dna_num[j][sub_index[0]:sub_index[1]]\n",
    "                seq_dist[i,j] = 1 - scipy.spatial.distance.hamming(seq_i,seq_j)\n",
    "    seq_dist = seq_dist + seq_dist.T           \n",
    "    reordering = scipy.cluster.hierarchy.leaves_list(\n",
    "                scipy.cluster.hierarchy.linkage(seq_dist_strong_tail))\n",
    "\n",
    "    ## plot the sequences\n",
    "    plt.figure(figsize=(4,18))\n",
    "    im = plt.matshow(\n",
    "        sites_strong_dna[reordering], \n",
    "        cmap=cmap_rhee, fignum=False) \n",
    "    plt.xlim([70,150]);\n",
    "    plt.colorbar(im, fraction=0.046, pad=0.04)\n",
    "\n",
    "\n",
    "display(sites)\n",
    "#print(bf.is_bedframe(sites))\n",
    "heatmap(sites)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0dd1f91-805e-40d9-9c32-1534c3405f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Visualization of Findings through Heatmap for all Strands and Upstream and Downstream\n",
    "df_list = ['upstream_pos.bed', 'downstream_pos.bed', 'upstream_neg.bed', 'upstream_neg.bed']\n",
    "for df in df_list:\n",
    "    sites = df\n",
    "    heatmap(sites) "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python bioframe",
   "language": "python",
   "name": "bioframe"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
