{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "22fb3649-43f8-4147-864f-8585540d5e02",
   "metadata": {},
   "outputs": [],
   "source": [
    "import bioframe as bf\n",
    "import pandas as pd \n",
    "import numpy as np "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "97f0083c-1150-419b-99ca-eed49e3ab404",
   "metadata": {},
   "outputs": [],
   "source": [
    "#input_bed = Chip-Seq Peak data in a bed file format \n",
    "#scan_bed = output of scan results of original input_bed \n",
    "\n",
    "#Function that creates dataframes for upstream and downstream of each strand \n",
    "def motif_input(input_bed, scan_bed): \n",
    "    input_bed.columns.values[0] = 'chrom'\n",
    "    input_bed.columns.values[1] = 'start'\n",
    "    input_bed.columns.values[2] = 'end'\n",
    "    input_bed = input_bed.iloc[:,0:3]\n",
    "    #input_bed = input_bed.drop(columns=['name', 'score'])\n",
    "    scan_bed = scan_bed.drop(columns=['name', 'score'])\n",
    "    overlapping_intervals = bf.overlap(input_bed, scan_bed, how = 'inner', suffixes = ('_1', '_2'))\n",
    "   \n",
    "    #Splitting the Dataframes into Positive and Negative Strand\n",
    "    df_pos = overlapping_intervals[overlapping_intervals['strand_2'] == '+']\n",
    "    df_neg = overlapping_intervals[overlapping_intervals['strand_2'] == '-']\n",
    "    \n",
    "    #Sliding the Dataframe Downstream for the Positive Strand\n",
    "    df_downstream_1 = df_pos.copy()\n",
    "    df_downstream_1['start_1'] = df_downstream_1['start_1'].astype(int)\n",
    "    df_downstream_1['end_1'] = df_downstream_1['end_1'].astype(int)\n",
    "    df_downstream_1['start_1'] = df_downstream_1['end_1']\n",
    "    df_downstream_1['end_1'] = df_downstream_1['start_1'] + 200\n",
    "    df_downstream_1 = df_downstream_1.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_downstream_1.columns.values[0] = '#chr'\n",
    "    df_downstream_1.columns.values[1] = 'start'\n",
    "    df_downstream_1.columns.values[2] = 'end'\n",
    "    df_downstream_1.to_csv('downstream_pos.bed', sep ='\\t', index = False)\n",
    "    \n",
    "    #Sliding the Dataframe Upstream for the Positive Strand\n",
    "    df_upstream_1 = df_pos.copy()\n",
    "    df_upstream_1['start_1'] = df_upstream_1['start_1'].astype(int)\n",
    "    df_upstream_1['end_1'] = df_upstream_1['end_1'].astype(int)\n",
    "    df_upstream_1['end_1'] = df_upstream_1['start_1']\n",
    "    df_upstream_1['start_1'] = df_upstream_1['start_1'] - 200\n",
    "    df_upstream_1 = df_upstream_1.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_upstream_1.columns.values[0] = '#chr'\n",
    "    df_upstream_1.columns.values[1] = 'start'\n",
    "    df_upstream_1.columns.values[2] = 'end'\n",
    "    df_upstream_1.to_csv('upstream_pos.bed', sep ='\\t', index = False)\n",
    "    \n",
    "    #Sliding the Dataframe Upstream for the Negative Strand\n",
    "    df_upstream_0 = df_neg.copy()\n",
    "    df_upstream_0['start_1'] = df_upstream_0['start_1'].astype(int)\n",
    "    df_upstream_0['end_1'] = df_upstream_0['end_1'].astype(int)\n",
    "    df_upstream_0['end_1'] = df_upstream_0['start_1']\n",
    "    df_upstream_0['start_1'] = df_upstream_0['start_1'] - 200\n",
    "    df_upstream_0 = df_upstream_0.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_upstream_0.columns.values[0] = '#chr'\n",
    "    df_upstream_0.columns.values[1] = 'start'\n",
    "    df_upstream_0.columns.values[2] = 'end'\n",
    "    df_upstream_0.to_csv('upstream_neg.bed', sep ='\\t', index = False)\n",
    "    \n",
    "    #Sliding the Dataframe Downstream for the Negative Strand\n",
    "    df_downstream_0 = df_neg.copy()\n",
    "    df_downstream_0['start_1'] = df_downstream_0['start_1'].astype(int)\n",
    "    df_downstream_0['end_1'] = df_downstream_0['end_1'].astype(int)\n",
    "    df_downstream_0['start_1'] = df_downstream_0['end_1']\n",
    "    df_downstream_0['end_1'] = df_downstream_0['start_1'] + 200\n",
    "    df_downstream_0 = df_downstream_0.drop(columns=['chrom_2', 'start_2','end_2','strand_2'])\n",
    "    df_downstream_0.columns.values[0] = '#chr'\n",
    "    df_downstream_0.columns.values[1] = 'start'\n",
    "    df_downstream_0.columns.values[2] = 'end'\n",
    "    df_downstream_0.to_csv('downstream_neg.bed', sep ='\\t', index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "980bcff2-15b7-4cb3-8a78-8b907bd8c602",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom_1</th>\n",
       "      <th>start_1</th>\n",
       "      <th>end_1</th>\n",
       "      <th>chrom_2</th>\n",
       "      <th>start_2</th>\n",
       "      <th>end_2</th>\n",
       "      <th>strand_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr12</td>\n",
       "      <td>3390741</td>\n",
       "      <td>3391298</td>\n",
       "      <td>chr12</td>\n",
       "      <td>3391085</td>\n",
       "      <td>3391103</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>3550331</td>\n",
       "      <td>3550695</td>\n",
       "      <td>chr12</td>\n",
       "      <td>3550523</td>\n",
       "      <td>3550541</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr12</td>\n",
       "      <td>3781582</td>\n",
       "      <td>3781985</td>\n",
       "      <td>chr12</td>\n",
       "      <td>3781736</td>\n",
       "      <td>3781754</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr12</td>\n",
       "      <td>3791855</td>\n",
       "      <td>3792158</td>\n",
       "      <td>chr12</td>\n",
       "      <td>3791991</td>\n",
       "      <td>3792009</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr12</td>\n",
       "      <td>4107368</td>\n",
       "      <td>4107782</td>\n",
       "      <td>chr12</td>\n",
       "      <td>4107546</td>\n",
       "      <td>4107564</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3938</th>\n",
       "      <td>chr16</td>\n",
       "      <td>95009671</td>\n",
       "      <td>95010171</td>\n",
       "      <td>chr16</td>\n",
       "      <td>95009965</td>\n",
       "      <td>95009983</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3939</th>\n",
       "      <td>chr16</td>\n",
       "      <td>95837166</td>\n",
       "      <td>95837511</td>\n",
       "      <td>chr16</td>\n",
       "      <td>95837357</td>\n",
       "      <td>95837375</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3940</th>\n",
       "      <td>chr16</td>\n",
       "      <td>95880393</td>\n",
       "      <td>95880760</td>\n",
       "      <td>chr16</td>\n",
       "      <td>95880594</td>\n",
       "      <td>95880612</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3941</th>\n",
       "      <td>chr16</td>\n",
       "      <td>96366411</td>\n",
       "      <td>96367030</td>\n",
       "      <td>chr16</td>\n",
       "      <td>96366818</td>\n",
       "      <td>96366836</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3942</th>\n",
       "      <td>chr16</td>\n",
       "      <td>96566723</td>\n",
       "      <td>96567197</td>\n",
       "      <td>chr16</td>\n",
       "      <td>96566932</td>\n",
       "      <td>96566950</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3943 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom_1   start_1     end_1 chrom_2   start_2     end_2 strand_2\n",
       "0      chr12   3390741   3391298   chr12   3391085   3391103        -\n",
       "1      chr12   3550331   3550695   chr12   3550523   3550541        +\n",
       "2      chr12   3781582   3781985   chr12   3781736   3781754        +\n",
       "3      chr12   3791855   3792158   chr12   3791991   3792009        -\n",
       "4      chr12   4107368   4107782   chr12   4107546   4107564        +\n",
       "...      ...       ...       ...     ...       ...       ...      ...\n",
       "3938   chr16  95009671  95010171   chr16  95009965  95009983        +\n",
       "3939   chr16  95837166  95837511   chr16  95837357  95837375        +\n",
       "3940   chr16  95880393  95880760   chr16  95880594  95880612        +\n",
       "3941   chr16  96366411  96367030   chr16  96366818  96366836        -\n",
       "3942   chr16  96566723  96567197   chr16  96566932  96566950        -\n",
       "\n",
       "[3943 rows x 7 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "input_bed = pd.read_csv('/home1/chavali/Supp_Tables_CSV/Supplemental_Table_S5.csv', header = 2, comment='#')\n",
    "\n",
    "scan_bed = pd.read_csv('/home1/chavali/beagan_files/gimme.CTCFes2i/motif_scan_results/GimmeMotifs_1.matches.bed', header = 3, sep='\\t', comment='#', \n",
    "                         names=['chrom', 'start', 'end', 'name', 'score', 'strand'])\n",
    "\n",
    "motif_input(input_bed, scan_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ed7efbca-30a6-40b1-ade7-ffc3fa0059e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "from gimmemotifs.denovo import gimme_motifs\n",
    "from gimmemotifs.scanner import Scanner\n",
    "\n",
    "peaks = ['upstream_pos.bed', 'downstream_pos.bed', 'upstream_neg.bed', 'upstream_neg.bed']\n",
    "outdir = ['upstream_pos.bed', 'downstream_pos.bed', 'upstream_neg.bed', 'upstream_neg.bed']\n",
    "params = {\n",
    "    \"tools\": \"Homer,BioProspector,MEME\",\n",
    "    \"genome\": \"mm9\"\n",
    "    }\n",
    "\n",
    "for (x,y) in zip(peaks,outdir): \n",
    "    motifs = gimme_motifs(x, y, params=params)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python bioframe",
   "language": "python",
   "name": "bioframe"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
